File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution(object):
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        total=sum(nums)
        if total%2!=0: return False
        target=total//2
        dp=[False]*(target+1)
        dp[0]=True
        for n in nums:
            for i in range(len(dp)-1,n-1,-1):
                if dp[i]: continue
                if dp[i-n]: dp[i]=True
                if dp[-1]: return True

        return False

        

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.cpp
class Solution {
public:
    vector<vector<int>>dp;
    bool solve(vector<int>& nums,int sum,int i){
        if(sum==0) return true;
        if(i>=nums.size()) return false;
        if(sum<0) return false;
        if(dp[i][sum]!=-1) return dp[i][sum];
        bool take=solve(nums,sum-nums[i],i+1);
        bool skip=solve(nums,sum,i+1);
        return dp[i][sum]=take || skip;
    }
    bool canPartition(vector<int>& nums) {
        int total=accumulate(nums.begin(),nums.end(),0);
        if(total%2!=0) return false;
        int k=total/2;
        dp.assign(nums.size(),vector<int>(k+1,-1));
        return solve(nums,k,0);
    }
};